selfdrive/ui/sunnypilot/qt/offroad/settings/vehicle_panel.cc

```diff
@@ -20,11 +20,101 @@ VehiclePanel::VehiclePanel(QWidget *parent) : QFrame(parent) {
   platformSelector = new PlatformSelector();
   list->addItem(platformSelector);
 
+//   hkgtuningToggle = new ParamControl("HKGtuning", (add all toggle parameters here), this);
+
+  QObject::connect(uiState(), &UIState::offroadTransition, [=](bool offroad) {
+    isOnroad = !offroad;
+    updateCarToggles();
+  });
+
   ScrollViewSP *scroller = new ScrollViewSP(list, this);
   vlayout->addWidget(scroller);
 
   QObject::connect(uiState(), &UIState::offroadTransition, platformSelector, &PlatformSelector::refresh);
 
+  QObject::connect(platformSelector, &PlatformSelector::clicked, [=]() {
+    QTimer::singleShot(100, this, &VehiclePanel::updateCarToggles);
+  });
+
+  QTimer *timer = new QTimer(this);
+  QObject::connect(timer, &QTimer::timeout, this, &VehiclePanel::updateCarToggles);
+  timer->start(1000);
+
   main_layout->addWidget(vehicleScreen);
   main_layout->setCurrentWidget(vehicleScreen);
 }
+
+void VehiclePanel::showRebootPrompt() {
+  QString question = tr("Changing this setting requires a reboot to take effect.\n\n"
+                       "Reboot now or later?");
+
+  reboot_confirmed = ConfirmationDialog::confirm(question, tr("Reboot Now"), this);
+  if (reboot_confirmed && !reboot_completed) {
+    params.putBool("DoReboot", true);
+    reboot_completed = true;
+  }
+}
+
+VehiclePanel::ToggleState VehiclePanel::getToggleState(bool hasOpenpilotLong) const {
+  if (!hasOpenpilotLong) {
+    return ToggleState::DISABLED_LONGITUDINAL;
+  }
+  if (isOnroad) {
+    return ToggleState::DISABLED_DRIVING;
+  }
+  return ToggleState::ENABLED;
+}
+
+void VehiclePanel::updateToggleState(ParamControl* toggle, bool hasOpenpilotLong) {
+  static const QString LONGITUDINAL_MSG = tr("Enable openpilot longitudinal control first.");
+  static const QString DRIVING_MSG = tr("Cannot modify while driving. Please go offroad first.");
+
+  ToggleState state = getToggleState(hasOpenpilotLong);
+
+  switch (state) {
+    case ToggleState::ENABLED:
+      toggle->setEnabled(true);
+      toggle->setToolTip("");
+      break;
+    case ToggleState::DISABLED_LONGITUDINAL:
+      toggle->setEnabled(false);
+      toggle->setToolTip(LONGITUDINAL_MSG);
+      break;
+    case ToggleState::DISABLED_DRIVING:
+      toggle->setEnabled(false);
+      toggle->setToolTip(DRIVING_MSG);
+      break;
+  }
+}
+
+void VehiclePanel::updateCarToggles() {
+  //bool hasOpenpilotLong = params.getBool("ExperimentalLongitudinalEnabled");
+
+  QString platform_name = QString::fromStdString(params.get("CarPlatformName"));
+  //if (platform_name.isEmpty()) {
+  //  hkgtuningToggle->setVisible(false);
+   // return;
+  //} To hide the toggle if no platform is selected
+
+  QVariantMap platform_data = platformSelector->loadPlatformList()[platform_name];
```

Comment:
We should avoid calling this too often. I will expose another param to cache this field persistently in the other PR, so we don't need to invoke whenever user tries to interact with the toggles.